routing/router_test.go

```diff
@@ -652,6 +653,373 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
+			},
+		},
+	}
+
+	// Prepare sequence of errors returned.
+	type errorEdge struct {
+		channelID uint64
+		source    string
+	}
+	var errorSeq = []errorEdge{
+		{
+			source:    "d",
+			channelID: d2eChanID,
+		},
+		{
+			source:    "e",
+			channelID: e2fChanID,
+		},
+		{
+			source:    "b",
+			channelID: b2dChanID,
+		},
+	}
+
+	amt := lnwire.NewMSatFromSatoshis(400)
+	var payHash [32]byte
+	payment := LightningPayment{
+		Target:      ctx.aliases["f"],
+		Amount:      amt,
+		FeeLimit:    noFeeLimit,
+		PaymentHash: payHash,
+		RouteHints:  routeHints,
+	}
+
+	// Temporary channel failure onion errors will be returned to prune channels.
+	update := lnwire.ChannelUpdate{}
+	chanFailure := lnwire.FailTemporaryChannelFailure{}
+	chanFailure.Update = &update
+	htlcError := htlcswitch.ForwardingError{}
+	htlcError.FailureMessage = &chanFailure
+
+	// Mock SendToSwitch to return the sequence of errors.
+	ctx.router.cfg.SendToSwitch = func(chanID lnwire.ShortChannelID,
+		_ *lnwire.UpdateAddHTLC, _ *sphinx.Circuit) ([32]byte, error) {
+
+		if len(errorSeq) == 0 {
+			return [32]byte{}, nil
+		}
+
+		// Prepare the error
+		source := errorSeq[0].source
+		channelID := errorSeq[0].channelID
+		htlcError.ErrorSource = ctx.aliases[source]
+		update.ShortChannelID = lnwire.NewShortChanIDFromInt(channelID)
+		update.Timestamp = uint32(testTime.Add(time.Minute).Unix())
+
+		err = signErrChanUpdate(ctx.privKeys[source], &update)
+		if err != nil {
+			t.Fatalf("Failed to sign channel on behalf of %v error: %v ", source, err)
+		}
+
+		// Pop the already returned error.
+		errorSeq = errorSeq[1:]
+
+		return [32]byte{}, &htlcError
+	}
+
+	// processUpdate ensures that the funding transaction is not spent,
+	// before further processing an edge update. Toggel AssumeChannelValid
+	// to avoid dropping the channel update, as the utxo set query is
+	// not mocked in this test.
+	ctx.router.cfg.AssumeChannelValid = true
+
+	// Send payment
+	_, route, err := ctx.router.SendPayment(&payment)
+	if err != nil {
+		t.Fatalf("failed to send payment: %v %v", err, payment)
+	}
+
+	// Verify all errors in planned sequence were returned.
+	if len(errorSeq) > 0 {
+		t.Fatalf("not all errors returned: %v errors left in sequence",
+			len(errorSeq))
+	}
+
+	// Check route is: A->C->E->D->F.
+	var expectedSeq = []string{"c", "e", "d", "f"}
+	var actualSeq []string
+	for i := 0; i < len(route.Hops); i++ {
+		actualSeq = append(actualSeq,
+			getAliasFromPubKey(route.Hops[i].PubKeyBytes[:],
+				ctx.aliases),
+		)
+	}
+	if !reflect.DeepEqual(expectedSeq, actualSeq) {
+		t.Fatalf("incorrect route: expected %s got %s",
+			strings.Join(expectedSeq, "->"),
+			strings.Join(actualSeq, "->"))
+	}
+
+	// The route should have the updated fee.
+	if route.TotalFees != lnwire.MilliSatoshi(150) {
+		t.Fatalf("expected %v MSat total fee, "+
+			"instead got %v MSat", 150, route.TotalFees)
+	}
+
+	// Now switch directions and test again.
+	ctx.router.missionControl.ResetHistory()
```

Comment:
Instead of manually driving the test twice, it should be made table driven, such that the two scenarios can be run by the same logic.

That being said, this is an excellent test! 😀